skyportal/facility_apis/slack.py

```diff
@@ -0,0 +1,188 @@
+              username: SkyPortal username of requester
+        """
+
+        for filt in request.payload["observation_choices"]:
+            if filt not in request.instrument.to_dict()["filters"]:
+                raise ValueError(f"Improper observation_choice {filt}")
+
+        if (
+            request.payload["exposure_time"] <= 0
+            or request.payload["exposure_time"] >= 7200
+        ):
+            raise ValueError(
+                "Exposure time must be greater than 0 or less than 2 hours"
+            )
+
+        if request.payload["exposure_counts"] <= 0:
+            raise ValueError("Number of exposures must be greater than 0")
+
+        # The target of the observation
+        target = {
+            'name': request.obj.id,
+            'instrument_request': request.allocation.instrument.name,
+            'ra': "{:.5f}".format(request.obj.ra),
+            'dec': "{:.5f}".format(request.obj.dec),
+            'filters': ",".join(request.payload["observation_choices"]),
+            'exposure_time': request.payload["exposure_time"],
+            'exposure_counts': request.payload["exposure_counts"],
+            'username': request.requester.username,
+        }
+
+        c = SkyCoord(ra=request.obj.ra * u.degree, dec=request.obj.dec * u.degree)
+
+        app_url = get_app_base_url()
+        skyportal_url = f"{app_url}/source/{request.obj.id}"
+        text = (
+            f"Name: {target['name']}\n"
+            f"SkyPortal Link: {skyportal_url}\n"
+            f"RA / Dec: {c.ra.to_string(sep=':')} {c.dec.to_string(sep=':')} ({target['ra']} {target['dec']})\n"
+            f"Filters: {target['filters']}\n"
+            f"Exposure time: {target['exposure_time']}\n"
+            f"Exposure counts: {target['exposure_counts']}\n"
+            f"Username: {target['username']}"
+        )
+
+        return target, textwrap.dedent(text)
+
+
+class SLACKAPI(FollowUpAPI):
+
+    """An interface to SLACK operations."""
+
+    # subclasses *must* implement the method below
+    @staticmethod
+    def submit(request):
+
+        """Submit a follow-up request to SLACK.
+
+        Parameters
+        ----------
+        request: skyportal.models.FollowupRequest
+            The request to add to the queue and the SkyPortal database.
+        """
+
+        from ..models import FacilityTransaction, DBSession
+
+        req = SLACKRequest()
+        requestgroup, requesttext = req._build_payload(request)
+
+        altdata = request.allocation.altdata
+
+        if not altdata:
+            raise ValueError('Missing allocation information.')
+
+        slack_microservice_url = (
+            f'http://127.0.0.1:{cfg.get("slack.microservice_port", 64100)}'
+        )
+
+        data = json.dumps(
+            {
+                "url": f"{SLACK_URL}/{altdata['slack_workspace']}/{altdata['slack_channel']}/{altdata['slack_token']}",
+                "text": requesttext,
+            }
+        )
+
+        r = requests.post(
```

Comment:
Since these are used in `followup_requests.py` in `post`, the `submit` call has to be made async and awaited, otherwise this will block our handlers.

Your PR is not the one to introduce the issue, so I will file an issue and we can fix it separately.